core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -455,10 +455,14 @@ private Object processWithExceptionHandling(Expression expression, Object contex
                 return process(expression, context);
             }
             catch (RuntimeException e) {
-                // HACK
-                // Certain operations like 0 / 0 or likeExpression may throw exceptions.
-                // Wrap them a FunctionCall that will throw the exception if the expression is actually executed
-                return createFailureFunction(e, type(expression), ExpressionInterpreter.this.metadata);
+                if (optimize) {
+                    // HACK
+                    // Certain operations like 0 / 0 or likeExpression may throw exceptions.
+                    // Wrap them a FunctionCall that will throw the exception if the expression is actually executed
+                    return createFailureFunction(e, type(expression), ExpressionInterpreter.this.metadata);
+                }
+                // Do not suppress exceptions during expression evaluation.
+                throw e;
```

Comment:
I don't think this is enough. We need to update the evaluation of CASE, COALESCE, etc, to take into account lazy evaluation of CASE bodies.